### PR TITLE
Modify diagnose to query metrics

### PR DESCRIPTION
**Pull request by vthapar**
As part of diagnose firewall test we actually test underlying infra
for firewall ports being open or not. A better test is to actually
try and querying metrics from a non-gw node. This is must-have for
the metrics redesign in which we no longer open any firewall ports.

Refer github.com/submariner-io/enhancements/pull/128

Signed-off-by: Vishal Thapar <5137689+vthapar@users.noreply.github.com>

### CODE REVIEW

File: pkg/diagnose/deployment_metrics.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+	"github.com/submariner-io/subctl/pkg/cluster"
+)
+
+const (
+	gwCurlMetricsCommand = "curl -I -m 10 submariner-gateway-metrics.submariner-operator.svc.cluster.local:8080/metrics"
+	gnCurlMetricsCommand = "curl -I -m 10 submariner-globalnet-metrics.submariner-operator.svc.cluster.local:8081/metrics"
+)
+
+func CheckMetricsConfig(clusterInfo *cluster.Info, status reporter.Interface) bool {
+	status.Start("Checking if the metrics are accessible from non-gateway nodes")
```
Comment (tpantelis) [inline pkg/diagnose/deployment_metrics.go]: This start block will automatically be ended by the nested start block on L65 so if the nested block is ended with a failure then this one will erroneously report success. So I think we should remove this block and incorporate the text in the nested block as suggested below.

File: pkg/diagnose/deployment_metrics.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	singleNode, err := clusterInfo.HasSingleNode()
+	if err != nil {
+		status.Failure(err.Error())
+		return false
+	}
+
+	if singleNode {
+		status.Success(singleNodeMessage)
+		return true
+	}
```
Comment (tpantelis) [inline pkg/diagnose/deployment_metrics.go:57]: Move this into `checkComponentMetrics`.

File: pkg/diagnose/deployment_metrics.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+	if !result {
+		status.Failure("Failed to access metrics from non-gateway node")
+		return false
+	}
+
+	status.Success("Metrics accessible from non-gateway node")
+
+	return true
```
Comment (tpantelis) [inline pkg/diagnose/deployment_metrics.go]:
```suggestion
	return result
```

File: pkg/diagnose/deployment_metrics.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+func checkComponentMetrics(clusterInfo *cluster.Info, status reporter.Interface, component, command string) bool {
+	status.Start("Querying %s metrics", component)
```
Comment (tpantelis) [inline pkg/diagnose/deployment_metrics.go]:
```suggestion
	status.Start("Checking if %s metrics are accessible from non-gateway nodes", component)
```

File: pkg/diagnose/deployment_metrics.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+	"github.com/submariner-io/admiral/pkg/reporter"
+	"github.com/submariner-io/subctl/pkg/cluster"
+)
+
+const (
+	gwCurlMetricsCommand = "curl -I -m 10 submariner-gateway-metrics.submariner-operator.svc.cluster.local:8080/metrics"
+	gnCurlMetricsCommand = "curl -I -m 10 submariner-globalnet-metrics.submariner-operator.svc.cluster.local:8081/metrics"
+)
+
+func CheckMetricsConfig(clusterInfo *cluster.Info, status reporter.Interface) bool {
```
Comment (tpantelis) [inline pkg/diagnose/deployment_metrics.go]: This is only accessed within the package.

```suggestion
func checkMetricsConfig(clusterInfo *cluster.Info, status reporter.Interface) bool {
```

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: pkg/diagnose/deployment_metrics.go (hunk 1 above)
Comment (vthapar) [reply pkg/diagnose/deployment_metrics.go:34]: Seems to be working fine:

```
 ✓ Checking if the metrics are accessible from non-gateway nodes
 ✓ Querying gateway metrics 
 ✓ The gateway metrics are accessible
 ✗ Querying globalnet metrics 
 ✗ Unable to access globalnet metrics service in submariner-operator namespace   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (7) Failed to connect to submariner-globalnet-metrics.submariner-operator.svc.cluster.local port 8081 after 1 ms: Connection refused

 ✗ Failed to access metrics from non-gateway node
```

```
 ✓ Checking if the metrics are accessible from non-gateway nodes
 ✗ Querying gateway metrics 
 ✗ Unable to access gateway metrics service in submariner-operator namespace   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (7) Failed to connect to submariner-gateway-metrics.submariner-operator.svc.cluster.local port 8080 after 1 ms: Connection refused

 ✓ Querying globalnet metrics 
 ✓ The globalnet metrics are accessible
 ✗ Failed to access metrics from non-gateway node
```

File: pkg/diagnose/deployment_metrics.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+	}
+
+	return result
+}
+
+func checkComponentMetrics(clusterInfo *cluster.Info, status reporter.Interface, component, command string) bool {
+	status.Start("Checking if %s metrics are accessible from non-gateway nodes", component)
+	defer status.End()
+
+	if !skipSingleNodeCheck {
```
Comment (tpantelis) [inline pkg/diagnose/deployment_metrics.go]: It's better to avoid global variables. If we want to cache the flag we should do it in `cluster.Info`.

**Review by tpantelis — CHANGES_REQUESTED** [review]

**Review by tpantelis — APPROVED** [review]

**Review by skitt — APPROVED** [review]

**Pull request merged**